File: ncaa_scraper/ncaa_scraper/football_personnel_readiness.py
"""Build an auditable, research-only football personnel feature readiness file.

The primary football forecast intentionally remains score, venue and team
identity only.  This module joins the retained exact-team personnel releases
to the *upcoming forecast slate* so we can measure which recruiting and
returning-production fields are available before deciding whether they are
safe enough to fit as future model features.  It never changes a prediction
or registers a challenger forecast.
"""

from __future__ import annotations

import hashlib
import json
import math
import sqlite3
from datetime import datetime, timezone


PERSONNEL_FIELDS = (
    "talent_composite",
    "talent_rank",
    "blue_chip_ratio",
    "off_returning",
    "def_returning",
    "overall_returning",
)
PERSONNEL_DATASETS = ("team_talent", "returning_production")


def _number(value):
    try:
        parsed = float(value)
    except (TypeError, ValueError):
        return None
    return parsed if math.isfinite(parsed) else None


def _digest(value) -> str:
    return hashlib.sha256(
        json.dumps(value, sort_keys=True, separators=(",", ":")).encode()
    ).hexdigest()[:16]


def _safe_json(value):
    try:
        parsed = json.loads(value)
    except (TypeError, ValueError, json.JSONDecodeError):
        return {}
    return parsed if isinstance(parsed, dict) else {}


def _personnel_rows(conn: sqlite3.Connection, season: int):
    """Return one source row per exact team/dataset, preserving conflicts."""
    rows = {}
    for dataset in PERSONNEL_DATASETS:
        for row in conn.execute(
            "SELECT team_id,record_key,stats_json FROM football_stats "
            "WHERE dataset=? AND season=? ORDER BY team_id,record_key",
            (dataset, season),
        ):
            team_id = str(row[0]) if row[0] not in (None, "") else None
            if not team_id:
                continue
            payload = _safe_json(row[2])
            target = rows.setdefault((team_id, dataset), [])
            payload["_record_key"] = row[1]
            target.append(payload)
    return rows


def _receipt_map(conn: sqlite3.Connection, season: int):
    receipts = {}
    for row in conn.execute(
        "SELECT dataset,receipt_json FROM football_sources "
        "WHERE season=? AND dataset IN ('team_talent','returning_production')",
        (season,),
    ):
        payload = _safe_json(row[1])
        if payload.get("fetched_at") and payload.get("sha256"):
            receipts[row[0]] = {
                "dataset": row[0],
                "season": season,
                "fetched_at": payload["fetched_at"],
                "sha256": payload["sha256"],
            }
    return receipts


def _team_context(team_id: str, rows):
    values = {}
    source_datasets = []
    conflicts = []
    for dataset in PERSONNEL_DATASETS:
        payloads = rows.get((team_id, dataset), [])
        if not payloads:
            continue
        source_datasets.append(dataset)
        for field in PERSONNEL_FIELDS:
            candidates = {
                value
                for payload in payloads
                if (value := _number(payload.get(field))) is not None
            }
            if len(candidates) > 1:
                conflicts.append(field)
            elif candidates:
                values[field] = next(iter(candidates))
    context = {
        "team_id": team_id,
        "team": None,
        "available_fields": sorted(values),
        "source_datasets": sorted(source_datasets),
        "conflicting_fields": sorted(set(conflicts)),
    }
    context.update({field: values.get(field) for field in PERSONNEL_FIELDS})
    return context


def _status(home, away):
    if home["conflicting_fields"] or away["conflicting_fields"]:
        return "conflict"
    available = len(set(home["available_fields"]) & set(away["available_fields"]))
    if available == len(PERSONNEL_FIELDS):
        return "complete"
    if available:
        return "partial"
    return "unavailable"


def build(
    conn: sqlite3.Connection,
    upcoming: list[dict],
    target_season: int = 2026,
    *,
    primary_model_id: str | None = None,
):
    """Return exact-ID context coverage for each forecasted upcoming game."""
    rows = _personnel_rows(conn, target_season)
    receipts = _receipt_map(conn, target_season)
    games = []
    for game in upcoming:
        if not game.get("prediction"):
            continue
        home = _team_context(str(game["home_id"]), rows)
        away = _team_context(str(game["away_id"]), rows)
        games.append(
            {
                "game_id": str(game["id"]),
                "kickoff": game.get("kickoff"),
                "home_id": str(game["home_id"]),
                "away_id": str(game["away_id"]),
                "home_name": game.get("home_name"),
                "away_name": game.get("away_name"),
                "home_division": game.get("home_division"),
                "away_division": game.get("away_division"),
                "model_id": (game.get("prediction") or {}).get("model_id"),
                "status": _status(home, away),
                "home": home,
                "away": away,
            }
        )
    counts = {status: sum(game["status"] == status for game in games) for status in ("complete", "partial", "conflict", "unavailable")}
    field_counts = {
        field: sum(
            field in game[side]["available_fields"]
            for game in games
            for side in ("home", "away")
        )
        for field in PERSONNEL_FIELDS
    }
    payload = {
        "version": "football-personnel-readiness-v1",
        "generated_at": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
        "target_season": target_season,
        "primary_model_id": primary_model_id,
        "scope": "Forecasted upcoming games only; exact team IDs from the retained personnel releases.",
        "coverage": {
            "forecast_games": len(games),
            "team_sides": len(games) * 2,
            "complete_games": counts["complete"],
            "partial_games": counts["partial"],
            "conflict_games": counts["conflict"],
            "unavailable_games": counts["unavailable"],
            "field_side_counts": field_counts,
            "personnel_teams": len({team_id for team_id, _ in rows}),
        },
        "feature_fields": list(PERSONNEL_FIELDS),
        "source_receipts": [receipts[key] for key in PERSONNEL_DATASETS if key in receipts],
        "limitations": [
            "This is feature readiness evidence, not a forecast and not a player availability ruling.",
            "Recruiting and returning-production fields remain outside the primary football model until leakage-safe dated validation is complete.",
            "Missing records do not imply a team has no talent or returning production; conflicting exact-team rows remain flagged.",
            "Team-level context does not establish eligibility, depth-chart order, injuries, transfers or expected snaps.",
        ],
        "games": games,
    }
    payload["id"] = "football-personnel-readiness-v1-" + _digest(payload)
    return payload

File: frontend/app/_lib/football-personnel-readiness.ts
import type { Game, FootballPersonnelReadiness } from "./data";

export type FootballPersonnelReadinessGame = FootballPersonnelReadiness["games"][number];

/** The fields in the research release, in the order used on matchup cards. */
const DEFINITIONS = [
  { key: "talent_composite", label: "Talent composite", format: "number" as const, direction: "higher" as const },
  { key: "talent_rank", label: "Talent rank", format: "rank" as const, direction: "lower" as const },
  { key: "blue_chip_ratio", label: "Blue-chip ratio", format: "percent" as const, direction: "higher" as const },
  { key: "overall_returning", label: "Overall returning", format: "percent" as const, direction: "higher" as const },
  { key: "off_returning", label: "Offense returning", format: "percent" as const, direction: "higher" as const },
  { key: "def_returning", label: "Defense returning", format: "percent" as const, direction: "higher" as const },
] as const;

export type FootballPersonnelReadinessRow = {
  key: (typeof DEFINITIONS)[number]["key"];
  label: string;
  format: "number" | "percent" | "rank";
  direction: "higher" | "lower";
  away: number | null;
  home: number | null;
  edge: "away" | "home" | "even" | "unavailable";
};

export type FootballPersonnelDivisionCoverage = {
  division: "d1" | "d2" | "d3";
  scheduled: number;
  complete: number;
  partial: number;
  conflict: number;
  unavailable: number;
  missing: number;
};

const finite = (value: unknown): number | null =>
  typeof value === "number" && Number.isFinite(value) ? value : null;

/** Join only by game ID and both exact team IDs; a mismatched row is withheld. */
export function personnelReadinessForGame(
  games: FootballPersonnelReadinessGame[],
  game: Pick<Game, "id" | "home_id" | "away_id">,
): FootballPersonnelReadinessGame | null {
  const row = games.find((candidate) => candidate.game_id === game.id);
  if (!row || row.home_id !== game.home_id || row.away_id !== game.away_id) return null;
  return row;
}

export function footballPersonnelReadinessRows(
  game: FootballPersonnelReadinessGame,
): FootballPersonnelReadinessRow[] {
  return DEFINITIONS.map(({ key, label, format, direction }) => {
    const away = finite(game.away[key]);
    const home = finite(game.home[key]);
    let edge: FootballPersonnelReadinessRow["edge"] = "unavailable";
    if (away != null && home != null) {
      if (away === home) edge = "even";
      else if (direction === "higher") edge = home > away ? "home" : "away";
      else edge = home < away ? "home" : "away";
    }
    return { key, label, format, direction, away, home, edge };
  });
}

export function personnelReadinessStatusLabel(status: FootballPersonnelReadinessGame["status"]): string {
  switch (status) {
    case "complete": return "Complete context";
    case "partial": return "Partial context";
    case "conflict": return "Conflicting source rows";
    case "unavailable": return "Context unavailable";
  }
}

function normalizedDivision(value: unknown): FootballPersonnelDivisionCoverage["division"] | null {
  if (typeof value !== "string") return null;
  const normalized = value.trim().toLowerCase().replace(/[^a-z0-9]/g, "");
  if (["fbs", "fcs", "d1", "division1", "divisioni"].includes(normalized)) return "d1";
  if (["d2", "division2", "divisionii"].includes(normalized)) return "d2";
  if (["d3", "division3", "divisioniii"].includes(normalized)) return "d3";
  return null;
}

/**
 * Reconcile personnel context to the upcoming slate using the same exact
 * game and team IDs as the matchup cards. A missing row remains missing; it
 * is never counted as unavailable context from a source response.
 */
export function footballPersonnelCoverageByDivision(
  games: Array<Pick<Game, "id" | "home_id" | "away_id" | "home_division" | "away_division">>,
  readinessGames: FootballPersonnelReadinessGame[],
): FootballPersonnelDivisionCoverage[] {
  const rows: FootballPersonnelDivisionCoverage[] = (["d1", "d2", "d3"] as const).map((division) => ({
    division,
    scheduled: 0,
    complete: 0,
    partial: 0,
    conflict: 0,
    unavailable: 0,
    missing: 0,
  }));
  const byDivision = new Map(rows.map((row) => [row.division, row]));
  for (const game of games) {
    const homeDivision = normalizedDivision(game.home_division);
    const awayDivision = normalizedDivision(game.away_division);
    const row = homeDivision && homeDivision === awayDivision ? byDivision.get(homeDivision) : undefined;
    if (!row) continue;
    row.scheduled += 1;
    const readiness = personnelReadinessForGame(readinessGames, game);
    if (!readiness) {
      row.missing += 1;
      continue;
    }
    row[readiness.status] += 1;
  }
  return rows;
}
